File: api/src/main/java/net/blueshell/api/permission/EventPermission.java
package net.blueshell.api.permission;

import net.blueshell.api.base.BasePermissionEvaluator;
import net.blueshell.api.common.enums.Role;
import net.blueshell.api.model.event.Event;
import net.blueshell.api.service.event.EventService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Component;

@Component
public class EventPermission extends BasePermissionEvaluator<Event, EventService> {

    @Autowired
    public EventPermission(EventService service) {
        super(service);
    }

    @Override
    public boolean hasPermission(Authentication authentication, Object targetDomainObject, String permission) {
        if (authentication == null || targetDomainObject == null || permission == null) {
            return false;
        }
        Event event = (Event) targetDomainObject;
        var principal = getPrincipal();
        return switch (permission) {
            case "read" -> event.isApproved() || event.getCommittee().hasMember(principal);
            case "write" -> event.getCommittee().hasMember(principal);
            case "signUp" -> event.isApproved() && (!event.isMembersOnly() || hasAuthority(Role.MEMBER));
            default -> false;
        };
    }

    @Override
    public boolean hasPermissionId(Authentication authentication, Object targetId, String permission) {
        if (authentication == null || targetId == null || permission == null) {
            return false;
        }
        Event event = service.findById((Long) targetId);
        return event != null && hasPermission(authentication, event, permission);
    }
}

File: api/src/main/kotlin/net/blueshell/api/permission/EventPermission.kt
package net.blueshell.api.permission

import net.blueshell.api.base.BasePermissionEvaluator
import net.blueshell.api.common.enums.Role
import net.blueshell.api.model.event.Event
import net.blueshell.api.service.event.EventService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component

@Component
class EventPermission @Autowired constructor(service: EventService) :
    BasePermissionEvaluator<Event, EventService>(service) {
    override fun hasPermission(
        authentication: Authentication?,
        targetDomainObject: Any?,
        permission: String?
    ): Boolean {
        if (authentication == null || targetDomainObject == null || permission == null) {
            return false
        }
        val event = targetDomainObject as Event
        val principal = getPrincipal()
        return when (permission) {
            "read" -> event.approved || event.committee.hasMember(principal)
            "write" -> event.committee.hasMember(principal)
            "signUp" -> event.approved && (!event.membersOnly || hasAuthority(Role.MEMBER))
            else -> false
        }
    }

    override fun hasPermissionId(authentication: Authentication?, targetId: Any?, permission: String?): Boolean {
        if (authentication == null || targetId == null || permission == null) {
            return false
        }
        val event = service.findById(targetId as Long)
        return event != null && hasPermission(authentication, event, permission)
    }
}
